### ISSUE TITLE
 'A loop has been detected while registering schema types.' exception when overriding types in the service provider

### ISSUE THREAD

**Issue by tatarincev** [graphql-dotnet/graphql-dotnet#1592]
I'm using the `GraphQL 3.0.0-preview-1490` version.
## Description
In our project, we would like to extend the existing schema types via overriding with the derived types in the service resolver.  We return for one schema type `FootType`  the instance of derived type `FooType2`.  But seems it doesn't work.

The exception  `A loop has been detected while registering schema types.`  is thrown when attempting to call   `Schema.Initialize()` . One prerequisite to get this error the overridden type must have fields with the own type (self-reference).

After a little investigation I  found a probable cause of this problem:
In this method
https://github.com/graphql-dotnet/graphql-dotnet/blob/042345659a2561d12d99448dd64cf4e57597be30/src/GraphQL/Types/GraphTypesLookup.cs#L388-L392
the line `var foundType = this[namedType];`  is uses a resulting type object for comparison with passed. For example for base type `FooType` the value will contain object instance of derived type `FooType2` and this method will return null that will cause the loop exception.
https://github.com/graphql-dotnet/graphql-dotnet/blob/042345659a2561d12d99448dd64cf4e57597be30/src/GraphQL/Types/GraphTypesLookup.cs#L175-L185

I'm just wondering is it intended behavior, and we will continue to look for another solution for extensibility, or can this be considered a bug?

### Steps to reproduce
```C#
 public class Foo
        {
            public string Name { get; set; }
        }
 public class FooType : ObjectGraphType<Foo>
        {
            public FooType()
            {
                Name = "Foo";
                Field(x => x.Name);
                //This nested field cause 'A loop has been detected while registering schema types.' exception in GraphTypesLookup.AddTypeWithLoopCheck
                Field<FooType>("parent", resolve: (ctx) => new Foo());
            }
        }
 public class FooType2 : FooType
        {
        }
 public class RootQuery : ObjectGraphType<object>
        {
            public RootQuery()
            {
                Field<FooType>(
                    "foo",
                    resolve: context => new FooType() { }
                );
            }
        }
        public class MySchema : Schema
        {
            public MySchema(IServiceProvider provider)
                : base(provider)
            {
                Query = new RootQuery();
            }
        }
  var services = new SimpleContainer();
            //Create a derived type FooType2 for FooType in type resolving
            services.Register<FooType>(() => new FooType2());
            var schema = new MySchema(new SimpleContainerAdapter(services));
            
            schema.Initialize(); //throw the  'A loop has been detected while registering schema types.' exception
```

### Expected result

The method should run without exception.

### Actual result
```
A loop has been detected while registering schema types.
There was an attempt to re-register 'Tests.GraphQLTests+FooType' with instance of 'Tests.GraphQLTests+FooType2'.
```

**Comment by sungam3r** [comment]
I did not understand what you want to do, but definitely such an error message is expected. All types in the schema must have unique names. Previously, without this check, there were a number of situations where it was possible to create an incorrectly configured schema, which led to incomprehensible errors in runtime. See #1421 . Maybe what you want to achieve can be done in a different way.

**Comment by tatarincev** [comment]
Thanks for the quick response. 
Try to rephrase my question :)
We want to override/substitute existing scheme types for another, without changing the source code where the schema defined and intend to use for this the `IServiceProvider` that will return the instance of type `FooType2` instead of original `FooType`.

**Comment by sungam3r** [comment]
>  resolve: context => new FooType() { }

Are you sure?

**Comment by tatarincev** [comment]
I guided by the following, why I can't  `cheat` and replace the declared type `FooType` with derived `FooType2`  since the GraphQL schema is built during execution anyway. At first glance, this problem seems trivial because from the system perspective won't be the two different types `FooType` and `FooType2` at the same time. 

In the provided example I used the empty `FooType2`  only to articulate the problem, in the real world scenarios the `FooType2` would have some new fields these values can be taken from different sources.
If we will have such an extension point we are able to distribute the root schema as a binary library and every consumer would extend any of the existing schema types with new fields using this approach (runtime type replacement).

I got your points about  `unpredictable consequences` but again I continue to be confident what this approach should work without any serious consequences and does not contradict existing implementation.

**Comment by sungam3r** [comment]
To be able to replace graphq types, we need to change the schema traversal algorithm. I see no way to quickly solve the problem. It takes time to think.

**Comment by sungam3r** [comment]
@tatarincev 3 years have passed and we have released v7.3 last month. There are a lot of changes in design and new features along this time. I'm sure you managed to solve this issue. In any way I suggest to upgrade to the latest-and-greatest version of GraphQL.NET to get the best error handling and schema validation. Closing.
